subworkflows/plink.cwl

```diff
@@ -0,0 +1,110 @@
+cwlVersion: v1.2
+class: Workflow
+id: plink
+doc: |
+  Plink subworkflow. Loads a VCF then runs genome and sexchecks on it.
+requirements:
+- class: ScatterFeatureRequirement
+- class: MultipleInputFeatureRequirement
+- class: InlineJavascriptRequirement
+- class: StepInputExpressionRequirement
+inputs:
+  input_vcfs: { type: 'File[]' }
+  genome_target_sites: { type: 'File?', doc: "Target sites for identity-by-descent genome analysis." }
+  output_basename: {type: 'string', doc: "String value to use as basename for outputs"}
+  genome:
+    type:
+      - 'null'
+      - type: record
+        fields:
+          - name: "gz"
+            type: boolean?
+            doc: "output to be gzipped"
+          - name: "full"
+            type: boolean?
+            doc: "adds additional fields"
+          - name: "rel-check"
+            type: boolean?
+            doc: "removes pairs of samples with different FIDs"
+          - name: "unbounded"
+            type: boolean?
+            doc: "turns off clipping"
+          - name: "nudge"
+            type: boolean?
+            doc: "adjusts the final estimates"
+    doc: "invokes an IBS/IBD computation"
+
+  # Resource Requirements
+  plink_cpu: {type: 'int?', doc: "CPUs to allocate to plink"}
+  plink_ram: {type: 'int?', doc: "RAM in GB to allocate to plink"}
+
+outputs:
+  genome_out: {type: 'File?', outputSource: plink_process_binary/genome_out }
+
+steps:
+  bcftools_view:
+    run: ../tools/bcftools_view.cwl
+    when: $(inputs.targets_file_include != null)
+    scatter: [input_vcf]
+    hints:
+    - class: sbg:AWSInstanceType
+      value: c5.4xlarge
+    in:
+      input_vcf: input_vcfs
+      targets_file_include: genome_target_sites
+      output_filename:
+        valueFrom: "temp.vcf.gz##idx##temp.vcf.gz.tbi"
+      write_index:
+        valueFrom: $(1 == 1)
+      output_type:
+        valueFrom: "z"
+      cpu:
+        valueFrom: $(1)
+      ram:
+        valueFrom: $(2)
+    out: [output]
+
+  bcftools_merge:
+    run: ../tools/bcftools_merge.cwl
+    when: $(inputs.run_tool.length > 1)
+    in:
+      run_tool: input_vcfs
+      input_vcfs:
+        source: [bcftools_view/output, input_vcfs]
+        pickValue: first_non_null
+      output_filename: 
+        valueFrom: "merged.vcf.gz"
+      output_type:
+        valueFrom: "z"
+    out: [output]
+
+  plink_load_variant_file:
+    run: ../tools/plink_load_variant_file.cwl
+    in:
+      input_vcf:
+        source: [bcftools_merge/output, bcftools_view/output, input_vcfs]
+        valueFrom: |
+          $(self[0] != null ? self[0] : self[1] != null && self[1][0] != null ? self[1][0] : self[2][0])
```

Comment:
Sort of a nest, if-else one-liner going on here?

Reply:
Yeah this is a weird one that took a lot of time to get right. We have three inputs:
- bcftools_merge/output: the merged file
- bcftools_view/output: an optional array of files that have had target regions picked out of them
- input_vcfs: the input array

The logic boils down to this:
- If bcftools_merge/output is there use it. otherwise...
- If the bcftools_view/output is not null and the first value is not null, use that first value. otherwise...
- Use the first value from input_vcfs

The reason I'm doing the first value thing is because bcftools_merge only runs when there are more than 1 VCF inputs. So if the bcftools_merge value isn't there that means I only have one input VCF. From there I just decide whether it comes from bcftools_view or input_vcfs.

Also Cavatica has an overly aggressive compiler in that I have to check if the bcftools_view array is not null otherwise the javascript compiler fails.